File: python/src/dijkstra.py
"""Dijkstra's Algorithm Implementation

Finds the shortest path from a source node to all other nodes in a weighted graph
with non-negative edge weights.

Time Complexity: O((V + E) log V) with binary heap
Space Complexity: O(V)
"""

import heapq
from typing import Dict, List, Tuple, Optional, Set


def dijkstra(
    graph: Dict[str, Dict[str, float]], 
    start: str
) -> Tuple[Dict[str, float], Dict[str, Optional[str]]]:
    """
    Dijkstra's algorithm for finding shortest paths from a source node.
    
    Args:
        graph: Adjacency list representation {node: {neighbor: weight}}
        start: Starting node
        
    Returns:
        Tuple of (distances, predecessors)
        - distances: Dict mapping each node to its shortest distance from start
        - predecessors: Dict mapping each node to its predecessor in the shortest path
        
    Example:
        >>> graph = {
        ...     'A': {'B': 4, 'C': 2},
        ...     'B': {'C': 1, 'D': 5},
        ...     'C': {'D': 8, 'E': 10},
        ...     'D': {'E': 2},
        ...     'E': {}
        ... }
        >>> distances, predecessors = dijkstra(graph, 'A')
        >>> distances['E']
        11
    """
    # Initialize distances and predecessors
    distances: Dict[str, float] = {node: float('inf') for node in graph}
    distances[start] = 0
    predecessors: Dict[str, Optional[str]] = {node: None for node in graph}
    
    # Priority queue: (distance, node)
    pq: List[Tuple[float, str]] = [(0, start)]
    visited: Set[str] = set()
    
    while pq:
        current_distance, current_node = heapq.heappop(pq)
        
        # Skip if already visited
        if current_node in visited:
            continue
            
        visited.add(current_node)
        
        # Skip if we found a better path already
        if current_distance > distances[current_node]:
            continue
        
        # Check all neighbors
        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight
            
            # If we found a shorter path, update it
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                predecessors[neighbor] = current_node
                heapq.heappush(pq, (distance, neighbor))
    
    return distances, predecessors


def dijkstra_path(
    graph: Dict[str, Dict[str, float]], 
    start: str, 
    end: str
) -> Tuple[float, List[str]]:
    """
    Find the shortest path between two specific nodes.
    
    Args:
        graph: Adjacency list representation
        start: Starting node
        end: Target node
        
    Returns:
        Tuple of (distance, path)
        - distance: Shortest distance from start to end
        - path: List of nodes in the shortest path
        
    Raises:
        ValueError: If no path exists between start and end
        
    Example:
        >>> graph = {'A': {'B': 1}, 'B': {'C': 2}, 'C': {}}
        >>> distance, path = dijkstra_path(graph, 'A', 'C')
        >>> path
        ['A', 'B', 'C']
    """
    distances, predecessors = dijkstra(graph, start)
    
    # Check if path exists
    if distances[end] == float('inf'):
        raise ValueError(f"No path exists from {start} to {end}")
    
    # Reconstruct path
    path: List[str] = []
    current: Optional[str] = end
    
    while current is not None:
        path.append(current)
        current = predecessors[current]
    
    path.reverse()
    
    return distances[end], path


def dijkstra_early_exit(
    graph: Dict[str, Dict[str, float]], 
    start: str, 
    target: str
) -> Tuple[float, List[str]]:
    """
    Optimized version that stops as soon as the target is reached.
    
    Args:
        graph: Adjacency list representation
        start: Starting node
        target: Target node
        
    Returns:
        Tuple of (distance, path)
    """
    distances: Dict[str, float] = {node: float('inf') for node in graph}
    distances[start] = 0
    predecessors: Dict[str, Optional[str]] = {node: None for node in graph}
    
    pq: List[Tuple[float, str]] = [(0, start)]
    visited: Set[str] = set()
    
    while pq:
        current_distance, current_node = heapq.heappop(pq)
        
        # Early exit if we reached the target
        if current_node == target:
            break
        
        if current_node in visited:
            continue
            
        visited.add(current_node)
        
        if current_distance > distances[current_node]:
            continue
        
        for neighbor, weight in graph[current_node].items():
            distance = current_distance + weight
            
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                predecessors[neighbor] = current_node
                heapq.heappush(pq, (distance, neighbor))
    
    # Reconstruct path
    if distances[target] == float('inf'):
        raise ValueError(f"No path exists from {start} to {target}")
    
    path: List[str] = []
    current: Optional[str] = target
    
    while current is not None:
        path.append(current)
        current = predecessors[current]
    
    path.reverse()
    
    return distances[target], path

File: rust/src/dijkstra.rs
//! Dijkstra's Algorithm Implementation
//!
//! Finds the shortest path from a source node to all other nodes in a weighted graph
//! with non-negative edge weights.
//!
//! Time Complexity: O((V + E) log V) with binary heap
//! Space Complexity: O(V)

use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap, HashSet};
use std::hash::Hash;

/// Graph represented as an adjacency list
pub type Graph<'a, T> = HashMap<T, Vec<(T, f64)>>;

/// Result of Dijkstra's algorithm
#[derive(Debug, Clone)]
pub struct DijkstraResult<T: Clone> {
    pub distances: HashMap<T, f64>,
    pub predecessors: HashMap<T, Option<T>>,
}

/// Path result between two nodes
#[derive(Debug, Clone)]
pub struct PathResult<T: Clone> {
    pub distance: f64,
    pub path: Vec<T>,
}

/// Priority queue item for Dijkstra's algorithm
#[derive(Debug, Clone)]
struct PQItem<T> {
    node: T,
    distance: f64,
}

impl<T> PartialEq for PQItem<T> {
    fn eq(&self, other: &Self) -> bool {
        self.distance == other.distance
    }
}

impl<T> Eq for PQItem<T> {}

impl<T> PartialOrd for PQItem<T> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl<T> Ord for PQItem<T> {
    fn cmp(&self, other: &Self) -> Ordering {
        // Reverse ordering for min-heap
        other.distance.partial_cmp(&self.distance).unwrap_or(Ordering::Equal)
    }
}

/// Run Dijkstra's algorithm from a source node
///
/// # Arguments
///
/// * `graph` - Adjacency list representation of the graph
/// * `start` - Starting node
///
/// # Returns
///
/// `DijkstraResult` containing distances and predecessors
///
/// # Example
///
/// ```
/// use graph_algorithms::dijkstra::{dijkstra, Graph};
/// use std::collections::HashMap;
///
/// let mut graph: Graph<&str> = HashMap::new();
/// graph.insert("A", vec![("B", 4.0), ("C", 2.0)]);
/// graph.insert("B", vec![]);
/// graph.insert("C", vec![]);
///
/// let result = dijkstra(&graph, &"A");
/// assert_eq!(result.distances[&"A"], 0.0);
/// assert_eq!(result.distances[&"B"], 4.0);
/// ```
pub fn dijkstra<T>(graph: &Graph<T>, start: &T) -> DijkstraResult<T>
where
    T: Eq + Hash + Clone,
{
    let mut distances: HashMap<T, f64> = HashMap::new();
    let mut predecessors: HashMap<T, Option<T>> = HashMap::new();
    let mut visited: HashSet<T> = HashSet::new();

    // Initialize distances
    for node in graph.keys() {
        distances.insert(node.clone(), f64::INFINITY);
        predecessors.insert(node.clone(), None);
    }
    distances.insert(start.clone(), 0.0);

    // Priority queue
    let mut pq = BinaryHeap::new();
    pq.push(PQItem {
        node: start.clone(),
        distance: 0.0,
    });

    while let Some(PQItem { node: current, distance: current_distance }) = pq.pop() {
        // Skip if already visited
        if visited.contains(&current) {
            continue;
        }
        visited.insert(current.clone());

        // Skip if we found a better path already
        if current_distance > distances[&current] {
            continue;
        }

        // Check all neighbors
        if let Some(neighbors) = graph.get(&current) {
            for (neighbor, weight) in neighbors {
                let distance = current_distance + weight;

                // If we found a shorter path, update it
                if distance < distances[neighbor] {
                    distances.insert(neighbor.clone(), distance);
                    predecessors.insert(neighbor.clone(), Some(current.clone()));
                    pq.push(PQItem {
                        node: neighbor.clone(),
                        distance,
                    });
                }
            }
        }
    }

    DijkstraResult {
        distances,
        predecessors,
    }
}

/// Find the shortest path between two specific nodes
///
/// # Arguments
///
/// * `graph` - Adjacency list representation
/// * `start` - Starting node
/// * `end` - Target node
///
/// # Returns
///
/// `Result` containing `PathResult` or error if no path exists
pub fn dijkstra_path<T>(graph: &Graph<T>, start: &T, end: &T) -> Result<PathResult<T>, String>
where
    T: Eq + Hash + Clone,
{
    let result = dijkstra(graph, start);

    // Check if path exists
    if result.distances[end].is_infinite() {
        return Err(format!("No path exists from {:?} to {:?}", start, end));
    }

    // Reconstruct path
    let mut path = Vec::new();
    let mut current = Some(end.clone());

    while let Some(node) = current {
        path.push(node.clone());
        current = result.predecessors[&node].clone();
    }

    path.reverse();

    Ok(PathResult {
        distance: result.distances[end],
        path,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_simple_graph() {
        let mut graph: Graph<&str> = HashMap::new();
        graph.insert("A", vec![("B", 4.0), ("C", 2.0)]);
        graph.insert("B", vec![("C", 1.0), ("D", 5.0)]);
        graph.insert("C", vec![("D", 8.0)]);
        graph.insert("D", vec![]);

        let result = dijkstra(&graph, &"A");

        assert_eq!(result.distances[&"A"], 0.0);
        assert_eq!(result.distances[&"B"], 4.0);
        assert_eq!(result.distances[&"C"], 2.0);
        assert_eq!(result.distances[&"D"], 9.0);
    }

    #[test]
    fn test_path_finding() {
        let mut graph: Graph<&str> = HashMap::new();
        graph.insert("A", vec![("B", 1.0), ("C", 4.0)]);
        graph.insert("B", vec![("C", 2.0), ("D", 5.0)]);
        graph.insert("C", vec![("D", 1.0)]);
        graph.insert("D", vec![]);

        let result = dijkstra_path(&graph, &"A", &"D").unwrap();

        assert_eq!(result.distance, 4.0);
        assert_eq!(result.path, vec!["A", "B", "C", "D"]);
    }

    #[test]
    fn test_no_path() {
        let mut graph: Graph<&str> = HashMap::new();
        graph.insert("A", vec![("B", 1.0)]);
        graph.insert("B", vec![]);
        graph.insert("C", vec![]);

        let result = dijkstra_path(&graph, &"A", &"C");

        assert!(result.is_err());
    }
}
